Optimised framebuffer updates in correspondence with the requested changes

Vanisha1606 opened the pull request:
I have reviewed the requested changes and recommendations from PR #33 and then deep-dived into the codebase once more.
I have changed my approach to this issue a slight, and am adding a new commit to the PR.

This PR fixes the issue #3 

I have done a minimal patch that lets the run loop detect if the newly returned framebuffer is physically the same as the old one (i.e. == in OCaml), and then skip repainting in that case. We still get the usual frame-delay logic (so it won’t spin too fast), but if we  return the same framebuffer from oour tick function, Claudius will skip doing the texture update/draw calls.

- After each tick, we are comparing updated_buffer with prev_buffer by physical equality (!= in OCaml).
- If they are not the same object, we do the usual framebuffer_to_bigarray + render_texture calls.
- If they are the same object, we skip the texture update/draw.
- We still do Sdl.delay diff at the start of each loop iteration, so the loop will not spin faster than our target frame rate (60fps in this snippet).
- Event polling continues exactly as before, so key presses etc. still work.

IMO this is a better approach to the issue. Is this approach correct or there are other things that are needed to be taken into consideration? Thanks :))
@mdales @patricoferris

Vanisha1606 commented:
@mdales Are there any changes you would like me to do?

mdales commented:
> @mdales Are there any changes you would like me to do?

Hey there! Unfortunately I'm travelling today, I'll try get to look at it tomorrow!

Vanisha1606 commented:
Okay @mdales Thank you!

mdales commented:
This is great, but you need to do both! This is because of how arrays work in OCaml (and some other languages). 

When you update the array in OCaml, you change the contents, but you don't change anything about the reference to that data, which is what is stored in the array variable. We can play with `utop` to see this in action:

```ocaml
utop # let a = Array.init 10 (fun i -> i);;
val a : int array = [|0; 1; 2; 3; 4; 5; 6; 7; 8; 9|]

utop # let b = a;;
val b : int array = [|0; 1; 2; 3; 4; 5; 6; 7; 8; 9|]

utop # b == a;;
- : bool = true

utop # a.(0) <- 42;;
- : unit = ()

utop # a;;
- : int array = [|42; 1; 2; 3; 4; 5; 6; 7; 8; 9|]

utop # a == b;;
- : bool = true
```

What you can see here is that at the start `a` and `b` are the same array, and then I modify the contents of the array via `a`, but when I check, `a` and `b` are still the same, which is because in OCaml `a` and `b` are just references to the array, rather than the array themselves.

The same is true in Python also for instance:

``` python
>>> a = list(range(10))
>>> a
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
>>> b = a
>>> b
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
>>> a[0] = 42
>>> a
[42, 1, 2, 3, 4, 5, 6, 7, 8, 9]
>>> b
[42, 1, 2, 3, 4, 5, 6, 7, 8, 9]
```

This is why you need the dirty bit from your first appoach also, as there's two ways a tick function can update the screen:

1. It can create a new framebuffer and return that (which is caught by this PR)
2. It can modify the passed in framebuffer and return that (which was caught by your dirty bit approach)

Both are required :)

But the good news is that you now understand the right run loop to modify, which you've done here, and you know how to do the dirty bit approach, which you did originally. You need to just merge those two, and add:

* when you check are teh arrays equal, if they are you also have to check the dirty bit
* the original PR needed to set the dirty bit in more places

I hope that all makes sense! You're so close, and it's great to see in this PR you've dived into the run loop and figured out what its doing, which is awesome, as it's one of the tougher parts of Claudius, where it interacts with SDL.

Vanisha1606 commented:
Hello @mdales I have made modifications in base.ml, framebuffer.ml, and framebuffer.mli, and have merged both the approaches. Now the run loop in base.ml will repaint when either a new framebuffer is returned (different reference) or when the same framebuffer is returned but its dirty flag indicates that in‐place updates occurred. In framebuffer.ml, the type is changed from a plain 2D array to a record that holds the pixel data in a field (named “data”) along with a mutable dirty flag, and all functions that modify the buffer now set that flag. Finally, framebuffer.mli exposes these changes without any extra formatting changes. This set of changes fully addresses the issue as described.
Is this implementation now correct or do I need to make any more changes or take other things into consideration?
Thank Youu :))

mdales commented on src/framebuffer.ml:
```diff
@@ -102,6 +106,7 @@ let draw_line (x0 : int) (y0 : int) (x1 : int) (y1 : int) (col : int) (buffer :
   in loop x0 y0 initial_error
 
 let draw_polygon (points : (int * int) list) (col : int) (buffer : t) =
+  buffer.dirty <- true;
```
This is too soon to decide in `draw_polygon`, as if the point list is empty we don't modify the buffer. In this instance this function doesn't modify the buffer directly, but calls draw_line, so I'd let that make the buffer as dirty.

Or put another way, only mark the buffer dirty when you know you differently change the contents: even when we call draw_line, draw_line might also decide not to do any work.

mdales commented on src/framebuffer.ml:
```diff
@@ -197,6 +201,7 @@ let rightmost span =
   | Pair (_, x1) -> x1
 
 let filled_triangle (x0 : int) (y0 : int) (x1 : int) (y1 : int) (x2 : int) (y2 : int) (col : int) (buffer : t) =
+  buffer.dirty <- true;
```
Again, this feels too early: What if the triangle was too small to draw? Currently looking at the code I suspect we always draw one pixel if the triangle is implossibly small, but if we ever fixed that, then I imagine we'd get into trouble. I'd feel much better if this change was next to the lines that actually update the buffer.

mdales commented on src/framebuffer.ml:
```diff
@@ -354,6 +359,7 @@ let interpolate_strand (strand : ((int * int) * (int * int)) list) : (int * span
   )
 
 let filled_polygon (points : (int * int) list) (col : int) (buffer : t) =
+  buffer.dirty <- true;
```
If there are no points we don't update the buffer.

mdales commented on src/framebuffer.ml:
```diff
@@ -419,6 +425,7 @@ let filled_polygon (points : (int * int) list) (col : int) (buffer : t) =
 (* ----- *)
 
 let draw_char (x : int) (y : int) (f : Font.t) (c : char) (col : int) (buffer : t) : int =
+  buffer.dirty <- true;
```
Not necessary here I think, as this calls `pixel_write` and that should set the buffer.

mdales commented on src/framebuffer.ml:
```diff
@@ -445,6 +452,7 @@ let draw_char (x : int) (y : int) (f : Font.t) (c : char) (col : int) (buffer :
   )
 
 let draw_string (x : int) (y : int) (f : Font.t) (s : string) (col : int) (buffer : t) =
+  buffer.dirty <- true;
```
This doesn't modify the buffer, but calls something that does, so let that do the work.

mdales commented on src/framebuffer.mli:
```diff
@@ -159,3 +158,9 @@ val pixel_read: int -> int -> t -> int option
+
+val is_dirty: t -> bool
+val set_dirty: t -> unit
+val clear_dirty: t -> unit
```
Can you add odoc comments for each of these please?

mdales commented on src/framebuffer.ml:
```diff
@@ -1,34 +1,36 @@
     None
 
 let draw_circle (x : int) (y : int) (r : float) (col : int) (buffer : t) =
+  buffer.dirty <- true;
```
This calls `pixel_write`, which will set the dirty bit, so doesn't need to explicitly call it.

mdales commented on src/framebuffer.ml:
```diff
@@ -50,15 +52,16 @@ let draw_circle (x : int) (y : int) (r : float) (col : int) (buffer : t) =
   done
 
 let filled_circle (x : int) (y : int) (r : float) (col : int) (buffer : t) =
+  buffer.dirty <- true;
```
Are there any scenarios where I can call filled_circle and not change the buffer? What if I draw a circle that's off the screen? A circle that is 0 size? In those cases I'd not expect the dirty bit to be set, but this means it always will be.

mdales commented on src/framebuffer.ml:
```diff
@@ -72,15 +75,16 @@ let filled_circle (x : int) (y : int) (r : float) (col : int) (buffer : t) =
   done
 
 let draw_line (x0 : int) (y0 : int) (x1 : int) (y1 : int) (col : int) (buffer : t) =
+  buffer.dirty <- true;
```
Similar for filled circle :)

mdales commented on src/framebuffer.mli:
```diff
@@ -159,3 +158,9 @@ val pixel_read: int -> int -> t -> int option
+
+(** {2 Dirty-bit support} *)
+
+val is_dirty: t -> bool
```
I don't see any implementation for this - did you forget to commit it?

mdales commented on src/base.ml:
```diff
@@ -89,33 +89,40 @@ let run (title : string) (boot : boot_func option) (tick : tick_func) (s : Scree
+          | Error (`Msg e) -> Sdl.log "Boot error: %s" e
+          | Ok () -> ()
+        ) else if Framebuffer.is_dirty updated_buffer then (
+          framebuffer_to_bigarray s updated_buffer bitmap;
```
This technically works, but you end up duplicating an important bit of code here in your two branches, and that means there's a risk that we update one and miss the other if we ever change `framebuffer_to_bigarray` - in software engineering this is referred to as "DRY" or "Don't Repeat Yourself".

There is actually, I think, a bug here also potentially: Surely we need to call `clear_dirty` in both cases? As even the "fresh" buffer returned from tick will still be dirty?

mdales reviewed (changes requested):
Well done, you've taken what is quite a challenging ticket and got the method down to what I think is the right approach. I've made a lot of comments, but please don't despair! Most of them are the same, around where I think the dirty bit is set a bit eagerly.

That does make me think that perhaps the problem is that we should *always* use `pixel_write` to update the buffer, as that would have made your job a lot easier here. If I'd done that originally, you'd only need to set the dirty_bit there (and in the map functions), but not in every circle/line/char function. That would be one approach to fix things, but I think just rationalising what you have is also fine as per the comments. 

The other missing thing are test cases - we definitely need to have some test cases around ensuring the bit is set and cleared using the accessor functions you added, and we should modify the existing drawing tests also to check the bit has been set.

Code reviews can be frustrating, but I want to be clear I'm very pleased with how this is going: you've taken on what was probably the most challenging of the beginner tickets and come out on the other side with a working approach! That's super good - we now just need to do the tidying part and we can get this in!

Vanisha1606 commented:
Hello @mdales I have resolving these conflicts and have pushed a new commit into the PR
In the newer commit - 
- We only set dirty when pixels are actually changed.
- Higher-level drawing functions (like draw_polygon, draw_circle, etc.) no longer unconditionally set dirty themselves. Instead, they rely on pixel_write (or equivalent per-pixel loops) to do so if anything is actually in range and modified.
- All direct array writes (like row.(xi) <- col) have been replaced by calls to pixel_write, or by loops that call pixel_write, so out-of-bounds writes do nothing (and do not set the dirty bit).
-  Only pixel_write marks the buffer as dirty when an in‑bounds pixel is actually modified.
- In the base.ml file, the two separate branches that both called framebuffer_to_bigarray and render_texture are now merged into one single branch. If either the buffer reference changes (i.e. a new buffer is returned) or the buffer is marked dirty, we update the texture and then call Framebuffer.clear_dirty updated_buffer. This ensures we clear the dirty flag in all cases after rendering.
- The run loop now uses a single branch that checks whether the buffer changed (i.e. updated_buffer ≠ prev_buffer) or if it’s marked dirty. In that case, it calls framebuffer_to_bigarray and render_texture and then clears the dirty flag. This eliminates the duplicated code and ensures clear_dirty is always called when needed.
- Also added necessary comments in the framebuffer.mli file as asked per reviews.

@mdales are the changes in accordance to the reviews? Also I have made some formatting changes and other necessary changes for the passing of the failing test cases.  

The test cases are now running properly :)
bash
eval $(opam env)
dune build
dune run


 I have attached the screenshot for the same as well.  Thanks a lot for the reviews!!

Vanisha1606 commented:
@mdales I have made the changes based on your feedback. I wanted to ask if I can work on another issue alongside this one, while you review this one?? Would that be a good idea? Thank you

mdales commented:
Just an observertion @Vanisha1606 - there's a bunch of changes in this pull request that feel like formatting changes or style changes. In theory, a great ambition, but in practice it now means I have a lot more code to read through and review, which means more work I have to do before I can approve this PR.

In general on a large project we should run a tool like dune auto format so that the code is consistent in format, I'll make a ticket for me to do that at some point, but in a multi-collaborator project you can't just rewrite code because you feel it is better a certain way, as if everyone does that we end up with all the code being written.

I appreciate you did this with the best intentions, but in big projects it both causes people to get upset because you changed their code for no technical reason, and it just means the reviewer has to check every single line changed for mistakes, slowing everything down.

This time I'll go through it and flag if I disagree, but in future please don't to tidying up work like this. If you think something needs tidying make a ticket for it and do the work under that ticket.

Again, this is on me, I should have moved to the ocaml auto-formatting before we started Outreachy, but it's also something everyone need to learn that "just tidying" has a bigger cost than anticipated :)

mdales commented on src/framebuffer.ml:
```diff
@@ -77,53 +78,46 @@ let draw_line (x0 : int) (y0 : int) (x1 : int) (y1 : int) (col : int) (buffer :
-    | true -> ()
-    | false -> (
+    pixel_write x y col buffer;
+    if (x = x1) && (y = y1) then ()
```
You changed the type of equality check here. Can you explain why? 

(I'm not saying it's wrong, but if you change code you should be able to to say what was wrong before, and why you picked the new version)

mdales reviewed (changes requested):
Hi there. Apologies, but it took me half an hour to read through the first half of the file, just checking each line didn't have an accidental change on it so can you please simplify this PR just to have the bits you changed in it? In particular automated checks failed due to some of the changes.

I'm keen to get your work in, as the core change here is a good one, but it's causing me a lot of work having to read format changes, and I've not yet got to reviewing the logic of the change, just all these extra lines changed. Again, you had the best intent, but in big software projects that causes a lot of work.

It's also worth noting that this PR will conflict with #42 - and the formatting changes will make the job of updating that PR harder also.

Very sorry to have to push this back, but given how many PRs there are, I need to do so.

mdales commented:
I appreciate the need to work on something else so you can feel progress is being made, but let's just get this one in, and we'll get you onto something else. I super keen to get this one in, as it's such an important change in terms of stopping Claudius wasting power, and you have all the parts there. Let's get it over the line, we're so close.

For what it's worth, the PR is otherwise looking strong now - I like the move to using pixel_write in all instances, which simplifies the code a lot, and let you remove some bounds checks I saw.

Do you have an idea for what you'd like to work on next thematically?

Vanisha1606 commented:
@mdales Should I revert back to the previous commit or should I open a new PR all together? I am very sorry that you have to go through all code again because of this formatting issue from my side , I will keep this in my mind that I don't do formatting untill required.
I will send you the issue which I would like to work on next .

Vanisha1606 commented:
@mdales Also another thing i see the ci tests are failing but they were successful on my local setup. Could you please tell if there are other things to be taken into consideration as well?

mdales commented:
Whatever is easier for you in terms of how to undo the changes. But looking at your branch here, what I'd do is:

1. First create a backup! 

```shell
$ git checkout -b backup-my-work Vanisha1606-dev-fix/efficient-frame-skip
$ git checkout Vanisha1606-dev-fix/efficient-frame-skip
```

You should now have a back should anything go wrong!

2. All the extra code changes are just in that last commit, so we can "upstage" that commit by doing:

```shell
$ git reset HEAD~1
```

You should now find that your branch is back to the second command of your PR, but all the changes from that last change are still there, as if you'd not yet committed them.

3. Use `git add -p` to select which changes you want. Use this to add all the changes that impact the topic for this ticket, and don't commit the other bits for now. If there's a line where you've done some changes that are both necessary and unnecessary, just add that, but don't add areas where there are many lines of unnecessary changes. If `git add -p` mixes them, rather than pressing 'y' or 'n' you can use 's' to split that block up into smaller parts.

4. Now do `git commit` and add just those parts you specified with `git add -p` earlier. Do NOT do `git commit -a`

5. Now force update the branch on GitHub: `git push --force origin Vanisha1606-dev-fix/efficient-frame-skip`

You will still have a bunch of uncommitted changes on your branch, for now I'd just ask git to stash those away: `git stash` - that means git has put them on a scratch pad somewhere should you need them, but they're gone from your working directory.

Does any of that make sense?

mdales commented:
> @mdales Also another thing i see the ci tests are failing but they were successful on my local setup. Could you please tell if there are other things to be taken into consideration as well?

This is in part why I stopped, as this confused me given you'd built things locally. I wonder if the CI is using a different version of OCaml that doesn't like the syntax versus what you have. Which version are you running? CI is running OCaml 5.2 currently.

(note that this change is on a line that I think was tidied but not related to the ticket, so I Assume won't be a problem in general next time)

Vanisha1606 commented:
@mdales I am using the version 5.3 but as you said i don’t think that would be a problem for the next time as it was on a tidied up line

Vanisha1606 commented:
@mdales I am using the version 5.3 but as you said i don’t think that would be a problem for the next time as it was on a tidied up line

Vanisha1606 commented:
@mdales Thankyou for guiding me through the changes and yes it all makes sense to me .
Also I would like to work on "track control changes "issue #13 .

Vanisha1606 commented:
Hey @mdales I have now strictly followed the conversations that were to be resolved. I see that base.ml and framebuffer.mli reviews are already resolved by you :)) .  I have updated the framebuffer.ml file strictly according to your reviews now as well, most of it contained removing the buffer which I had already set right? However I was thinking for covering the test cases around ensuring the bit is set and cleared using the accessor functions I added and for moving to pixel_write in all instances which simplifies the code a lot and let us remove the bound check cases I can open another ticket and work on it and raise a PR for those as well if it’s all right?

Vanisha1606 commented:
Hey @mdales I have gone all the reviews fixed my code accordingly. Thanks a lot :)) I also made some minute changes in the code to fix build errors of this sort : -

```bash
File "src/framebuffer.ml", line 485, characters 4-10:
485 |   ) buffer;
          ^^^^^^
Error: This expression has type t but an expression was expected of type
         int array array


File "src/framebuffer.ml", line 1:    
Error: The implementation src/framebuffer.ml
       does not match the interface src/framebuffer.ml: 
       The value is_dirty is required but not provided
       File "src/framebuffer.mli", line 165, characters 0-23:
         Expected declaration
       The value set_dirty is required but not provided
       File "src/framebuffer.mli", line 168, characters 0-24:
         Expected declaration
       The value clear_dirty is required but not provided
       File "src/framebuffer.mli", line 171, characters 0-26:
         Expected declaration

```
![WhatsApp Image 2025-03-27 at 10 10 01 PM](https://github.com/user-attachments/assets/15395f3d-1662-4b96-a6ab-448e1ec18a1b)

Now the build errors are all gone :) and tests are also running successfully, also reviewed and made corrections you requested.

mdales commented:
Well done @Vanisha1606!

But there is one bit I asked for that you missed :) Can you update `test_primatives.ml` to assert the dirty bit has been set after something has been drawn (you'll need to remember to clear it beforehand as it starts set). That way the tests actually not just pass, but also do check that all the work you've done doesn't get accidentally broken by someone in the future.

Vanisha1606 commented:
@mdales I have added the test case in the test_primitivies.ml file as you have asked, and ran tests once more !! Is it okaay now? Thanks for the suggestion
![WhatsApp Image 2025-03-28 at 3 40 41 PM](https://github.com/user-attachments/assets/2b0348f8-6cad-4cb7-a715-38b05626e23c)

mdales commented:
This is great, but it doesn't check every path you've changed. 

* Can you add a case for map/mapi/map_inplace?
* Can you change the primitive test case to repeat the test for all the primitives?

Why do we add tests? For me, there's two reasons:

* We get confidence that what we've added does what we think it does
* It stops someone in the future accidentally breaking the code.

And it's the second reason that is the most liberating the more you write software. If frees you up to move quickly: if I don't have good test coverage every modification I make I need to check so much more manually, but if I have good test coverage I can make changes in confidence that if I accidentally break something tests will tell me (or whoever it is doing the changes). That's why I'm so keen on everyone adding tests.

Is it tedious? Is it less fun than making new things? Yes. But it's a lot less tedious than when you next make a change and people start asking why things have broken :) It's a small amount of effort now that saves a lot of effort in the future.

There some things you can do to speed up the process here: rather than add a new primitive test, you can add the dirty bit check to all the existing tests in test_primative.ml, rather than create new tests.

I appreciate you want to move on to another ticket, and you want a chance to work on new things, and I really want to merge this code, but if I do that then I'll just end up writing all these tests myself later when I go to do any changes in this area, as I won't be comfortable without those tests being there. Please stick with it!

mdales commented:
Another way to make it easier to write the tests without writing a new test case for each primative:

```ocaml
  let test_dirty_bit_render_primitive _ =
    let primatives = [
      Primitives.Pixel ({x = 5; y = 5}, 1);
      Primitives.Line ({x = 2; y = 2}, {x = 5; y = 5}, 1);
      Primitives.Circle ({x = 5; y = 5}, 4.0, 1);
    ] in
    List.iter (fun prim ->
      let fb = Framebuffer.init (10, 10) (fun _ _ -> 0) in
      Framebuffer.clear_dirty fb;
      assert_equal ~msg:"dirty bit should be false after clear" false (Framebuffer.is_dirty fb);
      Framebuffer.render fb [prim];
      assert_equal ~msg:"dirty bit should be set after render" true (Framebuffer.is_dirty fb)
    ) primatives
 ```
 
 You could build up a list like this (just making sure it has one of every primative in it)

mdales commented:
As motivation, I took your branch for a spin, along with the mouse code PR, and it was great to see it working! (See video on #15) When I'm drawing I can see higher CPU load than when I'm not, which means your code is doing what it should!

Vanisha1606 commented:
@mdales I have made changes in the test file accordingly, also needed to make a lil extra changes for the tests to run successfully.
![WhatsApp Image 2025-03-29 at 5 45 03 PM](https://github.com/user-attachments/assets/ea9ae04d-3d89-449e-8294-b7e04a94f637)

mdales reviewed (approved):
Thank you! I appreciate this was a tough ticket, but thanks for sticking with it. Having good test coverage is super important as we have more people working on the project. And this is a great change to have in, so thank you very much!
